diff --git a/src/Plugin/views/display/EntityBrowser.php b/src/Plugin/views/display/EntityBrowser.php
index 1be05cb..5c65a15 100644
--- a/src/Plugin/views/display/EntityBrowser.php
+++ b/src/Plugin/views/display/EntityBrowser.php
@@ -226,7 +226,7 @@ class EntityBrowser extends DisplayPluginBase implements TrustedCallbackInterfac
    * {@inheritdoc}
    */
   public static function trustedCallbacks() {
-    return ['postRender', 'elementPreRender'];
+    return ['preRenderAddFieldsetMarkup', 'postRender', 'elementPreRender'];
   }
 
 }
